File: SWIFT/Cap08/c08ex08.swift
// c08ex08.swift

import Foundation

let FIM = 10
var MATRIZ = [Int](repeating: 0, count: FIM)
var INICIO = 0

func Vazia() -> Bool {
    if (INICIO == 0) {
        return true
    } else {
        return false
    }
}

func Cheia() -> Bool {
    if (INICIO == FIM) {
        return true
    } else {
        return false
    }
}

func Adicionar(_ ELEMENTO: Int) -> Bool {
    if (Cheia()) {
        return false
    } else {
        MATRIZ[INICIO] = ELEMENTO
        INICIO += 1
        return true
    }
}

func Retirar(_ ELEMENTO: inout Int) -> Bool {
    if (Vazia()) {
        return false
    } else {
        ELEMENTO = MATRIZ[0]
        for i in 0..<INICIO - 1 {
            MATRIZ[i] = MATRIZ[i + 1]
        }
        INICIO -= 1
        return true
    }
}

func Entrada() {
    print("Entre com um elemento numerico: ", terminator: "")
    if let X = Int(readLine() ?? "") {
        print()
        if (Adicionar(X)) {
            print("Elemento \(X) inserido na posicao \(INICIO).")
        } else {
            print("Impossivel adicionar \(X) - fila lotada.")
        }
        print()
    } else {
        print("\nEntrada invalida.\n")
    }
}

func Saida() {
    var ELEMENTO: Int = 0
    if (Retirar(&ELEMENTO)) {
        print("Elemento \(ELEMENTO) retirado do inicio da fila.")
    } else {
        print("Impossivel retirar elemento - fila vazia.")
    }
    print()
}

func Atual() {
    if (!(Vazia())) {
        print("\(MATRIZ[0]) - primeiro elemento da fila.")
    } else {
        print("Impossivel apresentar - fila vazia.")
    }
    print()
}

func Exibicao() {
    if (!(Vazia())) {
        for i in 0..<INICIO {
            print("Posicao: \(String(format: "%2d", i+1)) = \(MATRIZ[i])")
        }
    } else {
        print("Impossivel apresentar - fila vazia.")
    }
    print()
}

func Criar() {
    INICIO = 0
    for i in 0..<FIM {
        MATRIZ[i] = 0
    }
}

var OPCAO = 0
while (OPCAO != 6) {
    print("PROGRAMA: FILA\n")
    print("[1] - Entrada")
    print("[2] - Saida")
    print("[3] - Apresentar 1o. da fila")
    print("[4] - Apresentar a fila")
    print("[5] - Nova fila")
    print("[6] - Sair\n")
    print("Escolha uma opcao: ", terminator: "")

    if let entrada = readLine(), let op = Int(entrada) {
        OPCAO = op
        print()
        if (OPCAO != 6) {
            switch (OPCAO) {
                case 1: Entrada()
                case 2: Saida()
                case 3: Atual()
                case 4: Exibicao()
                case 5: Criar()
                default:
                    print("Opcao invalida. Tente novamente.\n")
            }
        }
    } else {
        print("Entrada invalida. Tente novamente.\n")
    }
}

File: KOTLIN/Cap08/c08ex08.kt
// c08ex08.kt

const val FIM = 10
var MATRIZ = IntArray(FIM)
var INICIO = 0

fun Vazia(): Boolean {
    if (INICIO == 0) {
        return true
    } else {
        return false
    }
}

fun Cheia(): Boolean {
    if (INICIO == FIM) {
        return true
    } else {
        return false
    }
}

fun Adicionar(ELEMENTO: Int): Boolean {
    if (Cheia())
        return false
    else {
        MATRIZ[INICIO] = ELEMENTO
        INICIO++
        return true
    }
}

fun Retirar(ELEMENTO: IntArray): Boolean {
    if (Vazia())
        return false
    else {
        ELEMENTO[0] = MATRIZ[0]
        for (I in 0 until INICIO - 1) {
            MATRIZ[I] = MATRIZ[I + 1]
        }
        INICIO--
        return true
    }
}

fun Entrada() {
    var X: Int
    print("Entre com um elemento numerico: ")
    X = readLine()?.toInt() ?: 0
    println()
    if (Adicionar(X))
        println("Elemento $X inserido na posicao $INICIO")
    else
        println("Impossivel adicionar $X - fila lotada.")
    println()
}

fun Saida() {
    val ELEMENTO = intArrayOf(0)
    if (Retirar(ELEMENTO))
        println("Elemento ${ELEMENTO[0]} retirado do inicio da fila.")
    else
        println("Impossivel retirar - fila vazia.")
    println()
}

fun Atual() {
    if (!Vazia())
        println("${MATRIZ[0]} - primeiro elemento da fila.")
    else
        println("Impossivel apresentar - fila vazia.")
    println()
}

fun Exibicao() {
    if (!Vazia())
        for (I in 0 until INICIO)
            println("Posicao: %2d = ${MATRIZ[I]}".format(I + 1))
    else
        println("Impossivel apresentar - fila vazia.")
    println()
}

fun Criar() {
    INICIO = 0
    for (I in 0 until FIM)
        MATRIZ[I] = 0
}

fun main() {
    Criar()
    var OPCAO: Int
    do {
        println("PROGRAMA: FILA\n")
        println("[1] - Entrada")
        println("[2] - Saida")
        println("[3] - Apresentar 1o. da fila")
        println("[4] - Apresentar a fila")
        println("[5] - Nova fila")
        println("[6] - Sair\n")
        print("Escolha uma opcao: ")
        OPCAO = readLine()?.toInt() ?: 6
        println()
        if (OPCAO != 6) {
            when (OPCAO) {
                1 -> Entrada()
                2 -> Saida()
                3 -> Atual()
                4 -> Exibicao()
                5 -> Criar()
                else -> println("Opcao invalida. Tente novamente.\n")
            }
        }
    } while (OPCAO != 6)
}
